src/Utils/Hooks/useCollectorSignals.tsx

```diff
@@ -0,0 +1,134 @@
+import { useCollectorSignals_artworksConnection$key } from "__generated__/useCollectorSignals_artworksConnection.graphql"
+import { useCollectorSignals_me$key } from "__generated__/useCollectorSignals_me.graphql"
+import { extractNodes } from "Utils/extractNodes"
+
+interface ArtworkData {
+  internalID: string
+  isAcquireable?: boolean | null
+}
+
+interface PartnerOffer {
+  artworkId?: string | null
+  endAt?: string | null
+}
+
+interface SignalResult {
+  partnerOffer: { endAt?: string | null } | null
+}
+
+interface SignalResultForArtworks {
+  [artworkID: string]: SignalResult | undefined
+}
+
+// Define interfaces for clearer type distinction
+interface UseCollectorSignalsSingleArtwork {
+  artwork: useCollectorSignals_artwork$key
+  artworksConnection?: null
+  me?: useCollectorSignals_me$key | null
+}
+
+interface UseCollectorSignalsMultipleArtworks {
+  artworksConnection?: useCollectorSignals_artworksConnection$key | null
+  artwork?: null
+  me?: useCollectorSignals_me$key | null
+}
+
+export function useCollectorSignals(
+  props: UseCollectorSignalsSingleArtwork
+): SignalResult
+// eslint-disable-next-line no-redeclare
+export function useCollectorSignals(
+  props: UseCollectorSignalsMultipleArtworks
+): SignalResultForArtworks
+
+// eslint-disable-next-line no-redeclare
+export function useCollectorSignals({
+  artwork,
+  artworksConnection,
+  me,
+}: UseCollectorSignalsSingleArtwork | UseCollectorSignalsMultipleArtworks):
+  | SignalResultForArtworks
+  | SignalResult {
+  const signalsPartnerOffersEnabled = useFeatureFlag(
+    "emerald_signals-partner-offers"
+  )
+  const meData = useFragment(ME_FRAGMENT, me)
+  const artworkData = useFragment(ARTWORK_FRAGMENT, artwork)
+  const artworksConnectionData = useFragment(
+    ARTWORKS_CONNECTION_FRAGMENT,
+    artworksConnection
+  )
+
+  const partnerOffers: PartnerOffer[] = extractNodes(
+    meData?.partnerOffersConnection
+  )
+  const artworks: ArtworkData[] = extractNodes(artworksConnectionData)
+
+  const isSingleArtwork = !!artworkData
+  const normalizedArtworks: ArtworkData[] = isSingleArtwork
+    ? [artworkData]
+    : artworks
+
+  const collectorSignals = normalizedArtworks.reduce<
+    UseCollectorSignalsMultipleArtworks
+  >((acc, artwork) => {
+    const signals = {} as SignalResult
+
+    if (!artwork.isAcquireable) {
+      return acc
+    }
+
+    if (signalsPartnerOffersEnabled) {
+      const offer = partnerOffers.find(
+        offer => offer.artworkId === artwork.internalID
+      )
+      if (offer) {
+        signals.partnerOffer = offer
+      }
+    }
+
+    if (Object.keys(signals).length > 0) {
+      acc[artwork.internalID] = signals
+    }
+    return acc
+  }, {} as UseCollectorSignalsMultipleArtworks)
+
+  const result = isSingleArtwork
+    ? Object.values(collectorSignals)[0]
+    : collectorSignals
+  return result
+}
+
+const ME_FRAGMENT = graphql`
+  fragment useCollectorSignals_me on Me {
+    partnerOffersConnection(first: 100) {
```

Comment:
Everything looks nicer now, thanks for adding coverage too! 
Just this number of connections seems too much, I don't know the context but is this the right amount we want?